### PR TITLE
READY: F handle resources with statement

### LINKED ISSUE #4890: Migrating resources to use "with" statements
- [x] I have read CONTRIBUTING.rst

##### Tribler version/branch+revision: #####
V7.3.1-RC3 and devel
##### Operating system and version: #####
All

I'll start by saying I'm not a developer for the project, but a user.

I noticed in the codebase a number of files (I have counted 11 so far), a couple of sqlite3 connections, and a couple of socket connections, all closed manually and not handled (meaning they're not closed in a "try-finally" at the least) for a possibility of exceptions occurring while said connections are live.

Examples:

Sqlite3 connection:
https://github.com/Tribler/tribler/blob/1b250e16e5d76ff60fe801e22b2dab3bd6907b3b/Tribler/Core/Upgrade/db72_to_pony.py#L91

Socket:
https://github.com/Tribler/tribler/blob/1516936729be1dc00dfa8095810c15893cb6f51e/Tribler/Core/Utilities/network_utils.py#L89

File:
https://github.com/Tribler/tribler/blob/01dda9aedfd0de0860a38523715ecde2740371bf/Tribler/Core/Category/FamilyFilter.py#L26

I'll be happy to aid the project and pull request to migrate said resources to use a "with" statement (which is of course compatible with python 2.5+ and 3.0+).

### PR DESCRIPTION
**Pull request by Solomon1732**
Closes #4890 and replaces #4895

### CODE REVIEW

**Comment by ichorid** [comment]
ok to test

File: Tribler/Core/Utilities/network_utils.py [hunk 1]
```diff
@@ -82,20 +82,21 @@ def _test_port(family, sock_type, port):
     assert sock_type in (socket.SOCK_DGRAM, socket.SOCK_STREAM), "Invalid sock_type value %s" % sock_type
     assert 0 < port <= 65535, "Invalid port value %s" % port
 
-    s = None
     try:
-        s = socket.socket(family, sock_type)
-        if sock_type == socket.SOCK_STREAM:
-            s.setsockopt(socket.SOL_SOCKET, socket.SO_LINGER, struct.pack('ii', 1, 0))
-        s.bind(('', port))
+        # In py2 socket lacks the __exit__ dunder method needed for with statements,
```
Comment (ichorid) [inline Tribler/Core/Utilities/network_utils.py]: As we do not support py2 anymore, the comment is unnecessary. Just do the py3 thing.

File: Tribler/Test/GUI/FakeTriblerAPI/utils/network.py [hunk 2]
```diff
@@ -74,16 +75,17 @@ def _test_port(family, sock_type, port):
...
     assert 0 < port <= 65535, "Invalid port value %s" % port
 
-    s = None
     try:
-        s = socket.socket(family, sock_type)
-        if sock_type == socket.SOCK_STREAM:
-            s.setsockopt(socket.SOL_SOCKET, socket.SO_LINGER, struct.pack('ii', 1, 0))
-        s.bind(('', port))
+        # In py2 socket lacks the __exit__ dunder method needed for with statements,
+        # which is where contextli.closing comes in.
```
Comment (ichorid) [inline Tribler/Test/GUI/FakeTriblerAPI/utils/network.py]: Again, the comment is unnecessary, as we don't support py2 anymore.

Comment (ichorid) [reply Tribler/Test/GUI/FakeTriblerAPI/utils/network.py]: As is `contextlib` usage, of course.
Nice trick, though :+1:

File: Tribler/Core/Utilities/network_utils.py (hunk 1 above)
Comment (Solomon1732) [reply Tribler/Core/Utilities/network_utils.py]: Removed

File: Tribler/Test/GUI/FakeTriblerAPI/utils/network.py (hunk 2 above)
Comment (Solomon1732) [reply Tribler/Test/GUI/FakeTriblerAPI/utils/network.py]: Removed. And thanks 😄

**Comment by Solomon1732** [comment]
Some of the commits are conflict merges. I also rebased the branch in question, so some of the pull commits should be squashed.

**Comment by Solomon1732** [comment]
Trying to squash commits seems to create more problems locally than I can solve. I just tried for a couple of hours to no avail to squash commits. While I can squash the commits serially it would look subpar and messy. I can squash the last 2 commits as fixups so they won't show up

**Comment by Solomon1732** [comment]
Trying from scratch. Not closing this PR

**Comment by ichorid** [comment]
That is strange that you got into trouble with squashing. Typically that's a completely trivial and safe procedure. Use `rebase -i devel` to initiate the interactive rebase, and squash along the way

**Comment by Solomon1732** [comment]
I tried being smart and move stuff around. I think I solved the mess, however. Thanks for the tip 😄

**Comment by Solomon1732** [comment]
One of the exceptions in the windows test comes from this
```
test_multifile_torrent (Tribler.Test.Core.Libtorrent.test_libtorrent_download_impl.TestLibtorrentDownloadImpl) ...   File "c:\users\heteoma\appdata\local\programs\python\python37\lib\asyncio\events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64_py3\tribler\Tribler\Core\Libtorrent\__init__.py", line 32, in done_cb
    result_future.set_result(func(*args, **kwargs))
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64_py3\tribler\Tribler\Core\Libtorrent\LibtorrentDownloadImpl.py", line 1044, in apply_ip_filter
    self.handle.apply_ip_filter(enable)
  File "c:\users\heteoma\appdata\local\programs\python\python37\lib\asyncio\events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64_py3\tribler\Tribler\Core\Libtorrent\__init__.py", line 32, in done_cb
    result_future.set_result(func(*args, **kwargs))
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64_py3\tribler\Tribler\Core\Libtorrent\LibtorrentDownloadImpl.py", line 1044, in apply_ip_filter
    self.handle.apply_ip_filter(enable)
  File "c:\users\heteoma\appdata\local\programs\python\python37\lib\asyncio\events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64_py3\tribler\Tribler\Core\Libtorrent\__init__.py", line 32, in done_cb
    result_future.set_result(func(*args, **kwargs))
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64_py3\tribler\Tribler\Core\Libtorrent\LibtorrentDownloadImpl.py", line 1044, in apply_ip_filter
    self.handle.apply_ip_filter(enable)
  File "c:\users\heteoma\appdata\local\programs\python\python37\lib\asyncio\events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64_py3\tribler\Tribler\pyipv8\ipv8\taskmanager.py", line 88, in done_cb
    future.result()
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_Tests\PR_win64_py3\tribler\Tribler\Core\Libtorrent\LibtorrentDownloadImpl.py", line 773, in save_resume_data
    handle.save_resume_data()
```
https://jenkins-ci.tribler.org/job/GH_Tribler_PR_Tests/job/PR_win64_py3/544/artifact/output/test_runners_output/3.err/*view*/
It mentions files I did not touch.

The [second](https://jenkins-ci.tribler.org/job/GH_Tribler_PR_Tests/job/PR_win64_py3/544/artifact/output/test_runners_output/0.err/*view*/) set of exceptions comes from a file I did change, however I am puzzled as to why it holds the file and causes the exception
```python
    def test_cleanup_pony_experimental_db(self):
        # Assert True is returned for a garbled db and nothing is done with it
        garbled_db = os.path.join(self.session_base_dir, 'garbled.db')
        with open(garbled_db, 'w') as f:
            f.write("123")
        self.assertRaises(sqlite3.DatabaseError, cleanup_pony_experimental_db, garbled_db)
        self.assertTrue(os.path.exists(garbled_db))
```

Reading from the [official docs](https://docs.python.org/3/reference/compound_stmts.html#the-while-statement) I also can't see why it throws the exception.

The oddest thing about it is that all the other tests of every other files I handled succeeded, but this specific one fails. It seems to hold a file while another process from the same test unit holes the same file for some reason.

**Comment by Solomon1732** [comment]
As can be seen [here](https://github.com/Tribler/tribler/blob/devel/Tribler/Test/Core/Upgrade/test_db72_to_pony.py#L119), while I did indeed handle the file, I did not change the test.

As can be [seen in my fork](https://github.com/Solomon1732/tribler/blob/f_handle_resources_with_statement/Tribler/Test/Core/Upgrade/test_db72_to_pony.py#L118), I did not handle the function which raises the exceptions. It seems that it does not arise from my pull request.

**Comment by Solomon1732** [comment]
I might have narrowed it down
```python
# Removes the file
cleanup_pony_experimental_db(pony_db_bak)
# Checks if it exists. If not, it fails
self.assertTrue(os.path.exists(pony_db_bak))
```
I changed it to `assertFalse`
https://github.com/Solomon1732/tribler/blob/f_handle_resources_with_statement/Tribler/Test/Core/Upgrade/test_db72_to_pony.py#L143

**Comment by ichorid** [comment]
Well, the test is the test. It **must** be `True`, because the test checks that the original database is never deleted accidentally. So, the check should only pass if the DB file still **exists** after calling `cleanup_pony_experimental_db` on it.

```python
# Checks if it exists. If not, it fails
self.assertTrue(os.path.exists(pony_db_bak)
```

**Comment by Solomon1732** [comment]
Ah, ok. Reverted the change in any case.

How can I test the specific test locally, by the way? I don't need python 2, do I?

**Comment by ichorid** [comment]
You don't need Python2. Just run the test in IDE. Or do something like `nosetests test_file.py`

**Comment by Solomon1732** [comment]
Oh, ok. Thanks.

By the way, thanks to major help from seems that the reason the test fails is because the `__exit__` dunder method _doesn't_ close the db connection. It still needs to be closed manually (even the official docs [here](https://docs.python.org/3/library/sqlite3.html#using-the-connection-as-a-context-manager) close it manually).

Instead of doing that I'm thinking of creating a small utility function. This way the db can be used w/o needing to close it manually. What do you think?

As a side note the `with` statement with a db connection "merely" flushes the transaction to the disk. It doesn't close the connection.

@birb007 helped me behind the scenes and was the one of figured out where I goofed.

**Comment by ichorid** [comment]
BTW, you accidentally got some commits that were already committed by @egbertbouman. This happens if you incorrectly synchronized the upstream `devel` branch with your local branch by, e.g. doing `git pull devel` on top of your local `devel` changed by you. Git can be confusing sometimes, I know :face_with_head_bandage: 

The proper way to fix this stuff is:
(`origin` is the upstream repo)
```
git checkout origin/devel
git branch -f devel
git checkout devel
```
`git branch -f <branchname>` is my favourite command for fixing git problems now. It *forces* the branch HEAD pointer onto a commit, which is the easiest way to completely synchronize your local branch with some remote branch.

After that, checkout your feature branch and then do the interactive rebase: `git rebase -i devel` and select *only your commits*. And then force-push the branch.

Good luck! :four_leaf_clover:

**Comment by ichorid** [comment]
Well, I knew something fishy was going on with those `with` statements in sqlite bindings! That's why I had to close the connections explicitly.

**Comment by ichorid** [comment]
About the "small utility function", it depends on how it will look. I suggest you first finish this PR with explicit `connection.close()` stuff, and then do the function in another PR.

**Comment by Solomon1732** [comment]
Thank you for the help!

> About the "small utility function", it depends on how it will look. I suggest you first finish this PR with explicit `connection.close()` stuff, and then do the function in another PR.
On it 👍 Good thinking 😄

**Comment by ichorid** [comment]
Also, interactive rebase gives you an opportunity to rename your commits. For instance, renaming `sqline` to `sqlite` in commit message :wink:

**Comment by ichorid** [comment]
Also, we're planning to change the folders structure in a few days, so **all** unfinished PRs will become completely broken beyond repair and un-mergeable afterwards. So, no pressure :wink:

**Comment by Solomon1732** [comment]
Can this be retested please?

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
Congratulations! Tests pass! :tada: 
Now you only have to clean the PR from @egbertbouman's rogue commits...

**Comment by Solomon1732** [comment]
Yay! 🎉
How do I clean? Every time I try and rebase they reappear for some reason…

**Comment by Solomon1732** [comment]
Nevermind. Fixed the mess. Somehow…

**Comment by Solomon1732** [comment]
https://jenkins-ci.tribler.org/job/GH_Tribler_PR_Tests/job/PR_win64_py3/551/artifact/output/test_runners_output/3.err/*view*/
This is the failure. Please someone tell me this isn't because of me… X_X

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
That happens sometimes...

Anyways, thanks for contributing!

**Comment by Solomon1732** [comment]
Happy to help! 😄 
Thank goodness it wasn't me this time around XD

**Pull request merged**